# User Bias in Data Cleaning
For your homework assignment this week, we will explore how our treatment of our data can impact the quality of our results.

**Dataset:**
The data is a Salary Survey from AskAManager.org. It’s US-centric-ish but does allow for a range of country inputs.

A list of the corresponding survey questions can be found [here](https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html).

 

# Imports

In [1161]:
from collections import Counter

from nltk.tokenize import word_tokenize
import pandas as pd

In [1162]:
df = pd.read_csv('survey_data.csv', parse_dates=[0], header=None)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28109 entries, 0 to 28108
Data columns (total 18 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   0       28109 non-null  datetime64[ns]
 1   1       28109 non-null  object        
 2   2       28034 non-null  object        
 3   3       28108 non-null  object        
 4   4       7273 non-null   object        
 5   5       28109 non-null  object        
 6   6       20793 non-null  float64       
 7   7       28109 non-null  object        
 8   8       211 non-null    object        
 9   9       3047 non-null   object        
 10  10      28109 non-null  object        
 11  11      23074 non-null  object        
 12  12      28027 non-null  object        
 13  13      28109 non-null  object        
 14  14      28109 non-null  object        
 15  15      27886 non-null  object        
 16  16      27938 non-null  object        
 17  17      27932 non-null  object        
dtypes: dat

In [1163]:
df.head()

0      1                              2   \
0 2021-04-27 11:02:10  25-34   Education (Higher Education)   
1 2021-04-27 11:02:22  25-34              Computing or Tech   
2 2021-04-27 11:02:38  25-34  Accounting, Banking & Finance   
3 2021-04-27 11:02:41  25-34                     Nonprofits   
4 2021-04-27 11:02:42  25-34  Accounting, Banking & Finance   

                                         3    4       5       6    7    8   \
0        Research and Instruction Librarian  NaN  55,000     0.0  USD  NaN   
1  Change & Internal Communications Manager  NaN  54,600  4000.0  GBP  NaN   
2                      Marketing Specialist  NaN  34,000     NaN  USD  NaN   
3                           Program Manager  NaN  62,000  3000.0  USD  NaN   
4                        Accounting Manager  NaN  60,000  7000.0  USD  NaN   

    9               10              11           12            13  \
0  NaN   United States   Massachusetts       Boston     5-7 years   
1  NaN  United Kingdom             NaN    Cambridge  8 - 10 years   
2  NaN              US       Tennessee  Chattanooga   2 - 4 years   
3  NaN             USA       Wisconsin    Milwaukee  8 - 10 years   
4  NaN              US  South Carolina   Greenville  8 - 10 years   

            14               15          16     17  
0    5-7 years  Master's degree       Woman  White  
1    5-7 years   College degree  Non-binary  White  
2  2 - 4 years   College degree       Woman  White  
3    5-7 years   College degree       Woman  White  
4    5-7 years   College degree       Woman  White

# Assignment
Your goal for this assignment is to observe how your data treatment during the cleaning process can skew or bias the dataset.

Before diving right in, stop and read through the questions associated with the dataset. As you can see, they are either free-form text entries or categorical selections. Knowing this, perform some exploratory data analysis (EDA) to investigate the "state" of the dataset.

[Add as many code cell below here as needs]


**Question:** How would you describe the "state" of this dataset? Be specific and detailed in your answer. (Think paragraphs rather than sentences).

The state of the dataset is filthy. All columns but one are held as strings, but many can be categories and one should be numerical. Many of the missing values are from not requiring answers to questions. Some of this extra information offered in the not-requiring-answers questions may not be as useful as the main questions. Having some information about the question in the column names would be nice. Some columns have missing values that could be filled easily based on other columns. Columns with free-form entries will be hard to work with but still contain useful information. Luckily, all the questions that are required to submit the survey appear to have no missing values and many can be binned into categories. 

## The Plan

Now, it is time to plan how you will clean up the dataset. You **are not** allowed to use any machine learning technique to clean the data. (No SMOTE! No machine learning! Or anything like that!)

**Question:** Based on your EDA above, detail how you would clean up this dataset. 
Things to consider: (This is not an exhaustive list)
- Are there columns that can't be effectively cleaned? If so, why?
- Are there columns that genuinely won't have a data value?
- Does it make sense to segment the dataset based on specific columns when determining how to handle the missing values?
- Are outliers a factor in this dataset?

Remember preserving as much of the data as possible is the goal. That means dropping rows with a missing value somewhere might not be the best idea.

I,ll add column names respectively: ['timestamp', 'age', 'industry', 'title', 'title_context', 'salary', 'additional_comp', 'currency', 'currency_other', 'income_context', 'country', 'state', 'city', 'experience', 'field_experience', 'education', 'gender', 'race']

The 'context' columns will be difficult to clean up. They are mostly null and the free-form entries will be too difficult to parse to be worth it. Year and month should be extracted from the 'Timestamp' column. 

age, industry, currency, state, experience, field_experience, education, gender, race -> categorical

This will probably involve some manual cleaning and use of other columns to fill in gaps. 

Salary, additional_comp -> numerical

Additional_comp is tricky as it could be added to the salary column, but that being the target variable, it would be better to keep it separate.

Title can be parsed for common words like 'manager' or 'engineer' to create a new column.

In order for this to go into a model, all columns will need to be numerical. That will involve some kind of encoding for the categorical columns. One-hot encoding will hugely increase the number of columns, but it will be necessary to not introduce bias from ordinal encoding besides where that is appropriate.

As I go along in the notebook, I will add more detail to this plan.

## Implementation

Based on the plan the you described above, go ahead and clean up the dataset.

[Add as many code cell below here as needs]

In [1164]:
df.columns = ['timestamp', 'age', 'industry', 'title', 'title_context', 'salary', 'additional_comp', 'currency', 'currency_other', 'income_context', 'country', 'state', 'city', 'experience', 'field_experience', 'education', 'gender', 'race']
df

timestamp    age                       industry  \
0     2021-04-27 11:02:10  25-34   Education (Higher Education)   
1     2021-04-27 11:02:22  25-34              Computing or Tech   
2     2021-04-27 11:02:38  25-34  Accounting, Banking & Finance   
3     2021-04-27 11:02:41  25-34                     Nonprofits   
4     2021-04-27 11:02:42  25-34  Accounting, Banking & Finance   
...                   ...    ...                            ...   
28104 2025-01-03 16:03:10  18-24   Engineering or Manufacturing   
28105 2025-01-16 10:00:47  35-44                  Entertainment   
28106 2025-01-23 19:38:04  35-44                     Nonprofits   
28107 2025-01-27 21:02:19  45-54              Computing or Tech   
28108 2025-02-04 09:27:51  25-34              Computing or Tech   

                                                title title_context   salary  \
0                  Research and Instruction Librarian           NaN   55,000   
1            Change & Internal Communications Manager           NaN   54,600   
2                                Marketing Specialist           NaN   34,000   
3                                     Program Manager           NaN   62,000   
4                                  Accounting Manager           NaN   60,000   
...                                               ...           ...      ...   
28104                           Applications Engineer           NaN   100000   
28105  Product communication and education specialist           NaN  1000000   
28106                      Director of content design           NaN   132000   
28107                                             CMO           NaN   279000   
28108                                         Trainee           NaN    25300   

       additional_comp currency currency_other income_context  \
0                  0.0      USD            NaN            NaN   
1               4000.0      GBP            NaN            NaN   
2                  NaN      USD            NaN            NaN   
3               3000.0      USD            NaN            NaN   
4               7000.0      USD            NaN            NaN   
...                ...      ...            ...            ...   
28104          21000.0      USD            NaN            NaN   
28105              NaN    Other            Czk            NaN   
28106              NaN      USD            NaN            NaN   
28107              0.0      USD            NaN            NaN   
28108              NaN      EUR            NaN            NaN   

               country           state         city      experience  \
0        United States   Massachusetts       Boston       5-7 years   
1       United Kingdom             NaN    Cambridge    8 - 10 years   
2                   US       Tennessee  Chattanooga     2 - 4 years   
3                  USA       Wisconsin    Milwaukee    8 - 10 years   
4                   US  South Carolina   Greenville    8 - 10 years   
...                ...             ...          ...             ...   
28104              USA           Texas       Dallas  1 year or less   
28105  Česká republika             NaN       Prague   21 - 30 years   
28106              USA        Missouri  Kansas City   11 - 20 years   
28107              USA      Washington      seattle   21 - 30 years   
28108           Italia             NaN       Milano  1 year or less   

      field_experience        education      gender  \
0            5-7 years  Master's degree       Woman   
1            5-7 years   College degree  Non-binary   
2          2 - 4 years   College degree       Woman   
3            5-7 years   College degree       Woman   
4            5-7 years   College degree       Woman   
...                ...              ...         ...   
28104   1 year or less  Master's degree       Woman   
28105      2 - 4 years      High School       Woman   
28106    11 - 20 years  Master's degree       Woman   
28107    11 - 20 years  Master's degree       Woman   
28108   1 y

In [1165]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28109 entries, 0 to 28108
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   timestamp         28109 non-null  datetime64[ns]
 1   age               28109 non-null  object        
 2   industry          28034 non-null  object        
 3   title             28108 non-null  object        
 4   title_context     7273 non-null   object        
 5   salary            28109 non-null  object        
 6   additional_comp   20793 non-null  float64       
 7   currency          28109 non-null  object        
 8   currency_other    211 non-null    object        
 9   income_context    3047 non-null   object        
 10  country           28109 non-null  object        
 11  state             23074 non-null  object        
 12  city              28027 non-null  object        
 13  experience        28109 non-null  object        
 14  field_experience  2810

#### Time

More granular time data than year and month is probably not useful. So I will drop the timestamp column after extracting year and month.

In [1166]:
df['year'] = df['timestamp'].dt.year
df['month'] = df['timestamp'].dt.month
df.drop('timestamp', axis=1, inplace=True)
df

age                       industry  \
0      25-34   Education (Higher Education)   
1      25-34              Computing or Tech   
2      25-34  Accounting, Banking & Finance   
3      25-34                     Nonprofits   
4      25-34  Accounting, Banking & Finance   
...      ...                            ...   
28104  18-24   Engineering or Manufacturing   
28105  35-44                  Entertainment   
28106  35-44                     Nonprofits   
28107  45-54              Computing or Tech   
28108  25-34              Computing or Tech   

                                                title title_context   salary  \
0                  Research and Instruction Librarian           NaN   55,000   
1            Change & Internal Communications Manager           NaN   54,600   
2                                Marketing Specialist           NaN   34,000   
3                                     Program Manager           NaN   62,000   
4                                  Accounting Manager           NaN   60,000   
...                                               ...           ...      ...   
28104                           Applications Engineer           NaN   100000   
28105  Product communication and education specialist           NaN  1000000   
28106                      Director of content design           NaN   132000   
28107                                             CMO           NaN   279000   
28108                                         Trainee           NaN    25300   

       additional_comp currency currency_other income_context  \
0                  0.0      USD            NaN            NaN   
1               4000.0      GBP            NaN            NaN   
2                  NaN      USD            NaN            NaN   
3               3000.0      USD            NaN            NaN   
4               7000.0      USD            NaN            NaN   
...                ...      ...            ...            ...   
28104          21000.0      USD            NaN            NaN   
28105              NaN    Other            Czk            NaN   
28106              NaN      USD            NaN            NaN   
28107              0.0      USD            NaN            NaN   
28108              NaN      EUR            NaN            NaN   

               country           state         city      experience  \
0        United States   Massachusetts       Boston       5-7 years   
1       United Kingdom             NaN    Cambridge    8 - 10 years   
2                   US       Tennessee  Chattanooga     2 - 4 years   
3                  USA       Wisconsin    Milwaukee    8 - 10 years   
4                   US  South Carolina   Greenville    8 - 10 years   
...                ...             ...          ...             ...   
28104              USA           Texas       Dallas  1 year or less   
28105  Česká republika             NaN       Prague   21 - 30 years   
28106              USA        Missouri  Kansas City   11 - 20 years   
28107              USA      Washington      seattle   21 - 30 years   
28108           Italia             NaN       Milano  1 year or less   

      field_experience        education      gender  \
0            5-7 years  Master's degree       Woman   
1            5-7 years   College degree  Non-binary   
2          2 - 4 years   College degree       Woman   
3            5-7 years   College degree       Woman   
4            5-7 years   College degree       Woman   
...                ...              ...         ...   
28104   1 year or less  Master's degree       Woman   
28105      2 - 4 years      High School       Woman   
28106    11 - 20 years  Master's degree       Woman   
28107    11 - 20 years  Master's degree       Woman   
28108   1 year or less  Master's degree         Man   

                                     race  year  month  
0                                   White  2021      4  
1                                   White  2021      4  
2                

### Categorical Columns
These appear to be the easy ones to categorize. I will start with these.

In [1167]:
df['age'] = df['age'].astype('category')
df['industry'] = df['industry'].astype('category')
df['country'] = df['country'].astype('category')
df['state'] = df['state'].astype('category')
df['city'] = df['city'].astype('category')
df['experience'] = df['experience'].astype('category')
df['field_experience'] = df['field_experience'].astype('category')
df['education'] = df['education'].astype('category')
df['gender'] = df['gender'].astype('category')
df.dtypes

age                 category
industry            category
title                 object
title_context         object
salary                object
additional_comp      float64
currency              object
currency_other        object
income_context        object
country             category
state               category
city                category
experience          category
field_experience    category
education           category
gender              category
race                  object
year                   int32
month                  int32
dtype: object

Check to see how many categories in each column.

In [1168]:
categorical_columns = df.select_dtypes(['category']).columns
df[categorical_columns].nunique()

age                    7
industry            1220
country              383
state                137
city                4842
experience             8
field_experience       8
education              6
gender                 5
dtype: int64

#### City

City has way too many categories to be useful. I'll keep the top 10 cities and put the rest in an 'other' category.

In [1169]:
df.city.value_counts().head(10)

city
Boston           772
Chicago          752
New York         711
Seattle          691
London           576
New York City    502
San Francisco    495
Los Angeles      462
Portland         423
Toronto          416
Name: count, dtype: int64

In [1170]:
df.loc[~df.city.isin(df.city.value_counts().head(10).index), 'city'] = 'Other'
df['city'] = df.city.cat.remove_unused_categories()
df.city.value_counts()

city
Other            22309
Boston             772
Chicago            752
New York           711
Seattle            691
London             576
New York City      502
San Francisco      495
Los Angeles        462
Portland           423
Toronto            416
Name: count, dtype: int64

That's less than ideal, but city can be a big factor in salary, so I don't want to lose that information. All the nulls were also set to other taking care of the missing values.

Next up, industry. There is a set number of options in the survey, so I will use that to categorize the industry column and set the rest to other.

#### Industry

In [1171]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28109 entries, 0 to 28108
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               28109 non-null  category
 1   industry          28034 non-null  category
 2   title             28108 non-null  object  
 3   title_context     7273 non-null   object  
 4   salary            28109 non-null  object  
 5   additional_comp   20793 non-null  float64 
 6   currency          28109 non-null  object  
 7   currency_other    211 non-null    object  
 8   income_context    3047 non-null   object  
 9   country           28109 non-null  category
 10  state             23074 non-null  category
 11  city              28109 non-null  category
 12  experience        28109 non-null  category
 13  field_experience  28109 non-null  category
 14  education         27886 non-null  category
 15  gender            27938 non-null  category
 16  race              2793

In [1172]:
df.industry.value_counts().head(50)

industry
Computing or Tech                       4712
Education (Higher Education)            2466
Nonprofits                              2420
Health care                             1899
Government and Public Administration    1893
Accounting, Banking & Finance           1812
Engineering or Manufacturing            1700
Marketing, Advertising & PR             1136
Law                                     1097
Business or Consulting                   853
Education (Primary/Secondary)            838
Media & Digital                          774
Insurance                                532
Retail                                   505
Recruitment or HR                        460
Property or Construction                 387
Art & Design                             359
Utilities & Telecommunications           357
Transport or Logistics                   304
Sales                                    286
Social Work                              274
Hospitality & Events                     262
E

Law Enforcement & Security is the lowest of the offered options as far as I can see. On second thought, I will use all the options with more than 100 samples and put the rest in 'other'.

In [1173]:
df['industry'] = df.industry.cat.add_categories('Other')
df.loc[~df.industry.isin(df.industry.value_counts().head(24).index), 'industry'] = 'Other'
df['industry'] = df.industry.cat.remove_unused_categories()
df.industry.value_counts()

industry
Computing or Tech                       4712
Education (Higher Education)            2466
Nonprofits                              2420
Other                                   2391
Health care                             1899
Government and Public Administration    1893
Accounting, Banking & Finance           1812
Engineering or Manufacturing            1700
Marketing, Advertising & PR             1136
Law                                     1097
Business or Consulting                   853
Education (Primary/Secondary)            838
Media & Digital                          774
Insurance                                532
Retail                                   505
Recruitment or HR                        460
Property or Construction                 387
Art & Design                             359
Utilities & Telecommunications           357
Transport or Logistics                   304
Sales                                    286
Social Work                              274
H

Not bad. At least other isn't the top category. Missing values also set to other. I could go through and try to set some with more nuance based on the other columns, but I think this is good enough for now.

Onto country and state.

#### Country

In [1174]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28109 entries, 0 to 28108
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               28109 non-null  category
 1   industry          28109 non-null  category
 2   title             28108 non-null  object  
 3   title_context     7273 non-null   object  
 4   salary            28109 non-null  object  
 5   additional_comp   20793 non-null  float64 
 6   currency          28109 non-null  object  
 7   currency_other    211 non-null    object  
 8   income_context    3047 non-null   object  
 9   country           28109 non-null  category
 10  state             23074 non-null  category
 11  city              28109 non-null  category
 12  experience        28109 non-null  category
 13  field_experience  28109 non-null  category
 14  education         27886 non-null  category
 15  gender            27938 non-null  category
 16  race              2793

In [1175]:
df.country.value_counts().head(50)

country
United States                9004
USA                          7946
US                           2612
Canada                       1572
United States                 668
U.S.                          581
UK                            574
United Kingdom                548
USA                           468
Usa                           448
United States of America      430
Australia                     318
United states                 208
usa                           184
Germany                       174
England                       134
united states                 116
Us                            105
Ireland                       103
New Zealand                    95
Uk                             85
Canada                         76
Australia                      68
United Kingdom                 65
France                         65
U.S.A.                         46
Spain                          44
United States of America       44
Netherlands                    43
Scotla

Ah man, I really wish they would have used a dropdown for country. I'll have to do some cleaning here.

In [1176]:
df['country'] = df.country.str.lower()
df['country'] = df.country.str.replace(r'\W', '', regex=True)
us_strings = ['us', 'unitedstates', 'unitedstatesofamerica', 'usa', 'unitesstates', 'america', 'unitedsates', 'theunitedstates', 'unitedstated', 'unitedstate', 'unitedstateofamerica', 'unitedstares', 'unitedstatea']
df.loc[df.country.isin(us_strings), 'country'] = 'us'
df.country.value_counts().head(50)

country
us                23155
canada             1680
uk                  704
unitedkingdom       635
australia           389
germany             197
england             169
ireland             125
newzealand          123
france               68
netherlands          57
spain                49
scotland             46
sweden               41
switzerland          38
belgium              35
thenetherlands       31
japan                29
denmark              23
india                23
singapore            20
southafrica          19
austria              17
finland              16
israel               14
norway               14
italy                14
malaysia             13
brazil               11
philippines          10
poland               10
china                 9
mexico                7
thailand              6
nz                    6
englanduk             6
colombia              5
czechrepublic         5
argentina             5
greatbritain          5
pakistan              5
portugal

In [1177]:
uk_strings = ['uk', 'unitedkingdom', 'england', 'scotland', 'wales', 'englanduk', 'greatbritain']
df.loc[df.country.isin(uk_strings), 'country'] = 'uk'
df.country.value_counts().head(50)

country
us                 23155
canada              1680
uk                  1569
australia            389
germany              197
ireland              125
newzealand           123
france                68
netherlands           57
spain                 49
sweden                41
switzerland           38
belgium               35
thenetherlands        31
japan                 29
denmark               23
india                 23
singapore             20
southafrica           19
austria               17
finland               16
norway                14
italy                 14
israel                14
malaysia              13
brazil                11
philippines           10
poland                10
china                  9
mexico                 7
nz                     6
thailand               6
colombia               5
argentina              5
pakistan               5
portugal               5
taiwan                 5
southkorea             5
greece                 5
czechrepublic    

Okay, I'm done spending time on that. I'll just use the countries with more than 100 samples and put the rest in 'other'.

In [1178]:
df.loc[~df.country.isin(df.country.value_counts().head(8).index), 'country'] = 'Other'
df.country.value_counts().head(9)

country
us            23155
canada         1680
uk             1569
Other           803
australia       389
germany         197
ireland         125
newzealand      123
france           68
Name: count, dtype: int64

If they selected a state, they should be in the US.

In [1179]:
df.loc[~df.state.isnull(), 'country'] = 'us'
df['country'] = df.country.astype('category')
df.country.value_counts()

country
us            23244
canada         1679
uk             1569
Other           718
australia       389
germany         195
ireland         125
newzealand      123
france           67
Name: count, dtype: int64

In [1180]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28109 entries, 0 to 28108
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               28109 non-null  category
 1   industry          28109 non-null  category
 2   title             28108 non-null  object  
 3   title_context     7273 non-null   object  
 4   salary            28109 non-null  object  
 5   additional_comp   20793 non-null  float64 
 6   currency          28109 non-null  object  
 7   currency_other    211 non-null    object  
 8   income_context    3047 non-null   object  
 9   country           28109 non-null  category
 10  state             23074 non-null  category
 11  city              28109 non-null  category
 12  experience        28109 non-null  category
 13  field_experience  28109 non-null  category
 14  education         27886 non-null  category
 15  gender            27938 non-null  category
 16  race              2793

#### State

In [1181]:
df.state.value_counts().head(55)

state
California                        2611
New York                          2174
Massachusetts                     1522
Texas                             1269
Illinois                          1213
Washington                        1185
District of Columbia               983
Pennsylvania                       943
Virginia                           786
Minnesota                          723
Ohio                               655
Colorado                           632
Oregon                             626
North Carolina                     601
Maryland                           564
Georgia                            552
Michigan                           545
Florida                            522
Wisconsin                          469
New Jersey                         399
Missouri                           345
Indiana                            329
Arizona                            302
Tennessee                          286
Connecticut                        239
Utah               

I'll keep all the single states and put the rest to 'multi'. Samples with countries other than the US will be put in 'not_us'.

In [1182]:
df['state'] = df.state.cat.add_categories(['not_us', 'multi'])
df.loc[~(df.country == 'us'), 'state'] = 'not_us'
df.loc[~df.state.isin(df.state.value_counts().head(52).index), 'state'] = 'multi'
df['state'] = df.state.cat.remove_unused_categories()
df.state.value_counts().head(53)

state
not_us                  4865
California              2611
New York                2174
Massachusetts           1522
Texas                   1269
Illinois                1213
Washington              1185
District of Columbia     983
Pennsylvania             943
Virginia                 786
Minnesota                723
Ohio                     655
Colorado                 632
Oregon                   626
North Carolina           601
Maryland                 564
Georgia                  552
Michigan                 545
Florida                  522
Wisconsin                469
New Jersey               399
Missouri                 345
Indiana                  329
Arizona                  302
Tennessee                286
multi                    282
Connecticut              239
Utah                     208
Kentucky                 191
Iowa                     179
Kansas                   153
South Carolina           148
Louisiana                129
Maine                    126
Alabama 

In [1183]:
df[categorical_columns].nunique()

age                  7
industry            25
country              9
state               53
city                11
experience           8
field_experience     8
education            6
gender               5
dtype: int64

In [1184]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28109 entries, 0 to 28108
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               28109 non-null  category
 1   industry          28109 non-null  category
 2   title             28108 non-null  object  
 3   title_context     7273 non-null   object  
 4   salary            28109 non-null  object  
 5   additional_comp   20793 non-null  float64 
 6   currency          28109 non-null  object  
 7   currency_other    211 non-null    object  
 8   income_context    3047 non-null   object  
 9   country           28109 non-null  category
 10  state             28109 non-null  category
 11  city              28109 non-null  category
 12  experience        28109 non-null  category
 13  field_experience  28109 non-null  category
 14  education         27886 non-null  category
 15  gender            27938 non-null  category
 16  race              2793

### Salary

In [1185]:
df['salary'].str.replace(r'\W', '', regex=True).astype(float).describe()

count    2.810900e+04
mean     3.619204e+05
std      3.619273e+07
min      0.000000e+00
25%      5.400000e+04
50%      7.500000e+04
75%      1.097690e+05
max      6.000070e+09
Name: salary, dtype: float64

In [1186]:
df['num_sal'] = df['salary'].str.replace(r'\W', '', regex=True).astype(float)
df.loc[df.num_sal < 100]

age                       industry  \
97     55-64                          Other   
166    25-34                    Health care   
895    45-54  Education (Primary/Secondary)   
968    25-34                Media & Digital   
2229   25-34    Marketing, Advertising & PR   
...      ...                            ...   
28026  35-44                         Retail   
28049  25-34   Education (Higher Education)   
28093  45-54              Computing or Tech   
28096  25-34         Business or Consulting   
28098  18-24                          Other   

                                   title title_context salary  \
97                Quality Assurance Lead           NaN     58   
166               occupational therapist           NaN     35   
895            Special Education Teacher           NaN     38   
968                      Managing Editor           NaN     61   
2229   Project Manager / Account Manager           NaN     33   
...                                  ...           ...    ...   
28026                       CRM manager   CRM manager      70   
28049                               fasd        asdsda      5   
28093                     CYBER SECURITY            NO     40   
28096                     BI Consultant            NaN     10   
28098                      data engineer           NaN      0   

       additional_comp currency currency_other              income_context  \
97                 NaN      USD            NaN                         NaN   
166                NaN      EUR            NaN                         NaN   
895                0.0      USD            NaN                         NaN   
968                0.0      USD            NaN                         NaN   
2229               0.0      EUR            NaN  Before tax, insurance etc.   
...                ...      ...            ...                         ...   
28026              NaN      EUR            NaN                         NaN   
28049              NaN      EUR             ff                          ff   
28093           1200.0      EUR          other                          no   
28096              NaN    Other        Rupees                          NaN   
28098              NaN      USD            NaN                         NaN   

       country     state   city      experience field_experience  \
97          us  Maryland  Other   21 - 30 years     8 - 10 years   
166      Other    not_us  Other     2 - 4 years      2 - 4 years   
895         us     Idaho  Other   21 - 30 years    11 - 20 years   
968         us  Maryland  Other       5-7 years        5-7 years   
2229   germany    not_us  Other       5-7 years      2 - 4 years   
...        ...       ...    ...             ...              ...   
28026  germany    not_us  Other    8 - 10 years        5-7 years   
28049       us     multi  Other   21 - 30 years        5-7 years   
28093       us     multi  Other     2 - 4 years      2 - 4 years   
28096    Other    not_us  Other     2 - 4 years      2 - 4 years   
28098    Other    not_us  Other  1 year or less   1 year or less   

             education gender  \
97     Master's degree  Woman   
166     College degree  Woman   
895    Master's degree  Woman   
968     College degree  Woman   
2229   Master's degree  Woman   
...                ...    ...   
28026  Master's degree  Woman   
28049              NaN    NaN   
28093      High School    Man   
28096   College degree    Man   
28098  Master's degree    Man   

                                                    race  year  month  num_sal  
97                                                 White  2021      4     58.0  
166                                                White  2021      4     35.0  
895                                                White  2021      4     38.0  
968                                                White  2021      4     61.0  
2229                                               White  2021      4     33.0  
...                       

In [1187]:
df.num_sal.quantile([0.001, 0.999])

0.001    3.500000e+01
0.999    4.142720e+06
Name: num_sal, dtype: float64

I'm about at my wits end with this dataset. I'm going to drop the rows where salary is below the bottom 0.1% and top 99.9% as these are errors or outliers.

In [1188]:
df = df[df.num_sal >= df.num_sal.quantile(0.001)]
df = df[df.num_sal <= df.num_sal.quantile(0.999)]
df.num_sal.describe()

count    2.805200e+04
mean     9.239735e+04
std      1.011973e+05
min      3.500000e+01
25%      5.400000e+04
50%      7.500000e+04
75%      1.092000e+05
max      4.000000e+06
Name: num_sal, dtype: float64

In [1189]:
df.loc[df.num_sal > 1000000]

age                              industry  \
603         45-54                           Health care   
1449        45-54                Business or Consulting   
2124        55-64                          Art & Design   
3162        25-34                           Health care   
3937        45-54                     Computing or Tech   
4206        25-34          Engineering or Manufacturing   
4499        35-44                     Computing or Tech   
5146        45-54                                 Other   
5306        35-44                     Computing or Tech   
5755        25-34                           Health care   
6783        25-34                     Computing or Tech   
9238        35-44  Government and Public Administration   
10508       25-34                                Retail   
10904       35-44          Engineering or Manufacturing   
10978       25-34                Business or Consulting   
11134       45-54                                   Law   
11414       25-34                                 Other   
11693       35-44                     Computing or Tech   
11913       25-34                     Computing or Tech   
14429       25-34                                 Other   
15509       35-44                     Computing or Tech   
15795       25-34                         Entertainment   
16165       35-44                                   Law   
16769       25-34           Marketing, Advertising & PR   
16982       35-44                     Computing or Tech   
17708       25-34                Business or Consulting   
18757       35-44                     Computing or Tech   
18832       35-44                     Computing or Tech   
18943       25-34                     Computing or Tech   
20344       35-44          Education (Higher Education)   
21320       25-34                     Computing or Tech   
21322       25-34         Accounting, Banking & Finance   
21995       35-44                                 Other   
23437       25-34                       Media & Digital   
24940       35-44         Education (Primary/Secondary)   
24955       25-34                           Health care   
26558       45-54                Business or Consulting   
26747       25-34                     Computing or Tech   
27630       25-34                     Computing or Tech   
27800       55-64                             Insurance   
27857       25-34                     Computing or Tech   
28043       25-34                     Computing or Tech   
28061       25-34                           Health care   
28084  65 or over                                 Other   
28094       18-24                     Computing or Tech   

                                                  title  \
603                                          Exec admin   
1449                                        Design Lead   
2124                                     Owner and CEO    
3162                                        IT Director   
3937                                     Senior Manager   
4206                                  Software Engineer   
4499                                           Producer   
5146                                  translator/editor   
5306                                   Network Engineer   
5755    Attending Physician (general internal medicine)   
6783                        Principal Software Engineer   
9238                              Senior Policy Advisor   
10508                      Compliance / Privacy Manager   
10904                              Executive Assistant    
10978                                      Lead Analyst   
11134                                           Partner   
11414                             Localization Director   
11693                                 Master Specialist   
11913                         Machine Learning Engineer   
14429                       Customer Service Specialist   
15509                                   Product Manager   
1

This is rediculous. If the target is at different scales the ML model is going to have a really hard time. I could go through and apply the conversion rates to the salary column, but that sounds super tedious. I'll just drop the rows with other currencies.

In [1190]:
df = df[df.currency == 'USD']
df = df[df.currency_other.isnull()]
df.drop(['currency', 'currency_other'], axis=1, inplace=True)
df

age                       industry  \
0      25-34   Education (Higher Education)   
2      25-34  Accounting, Banking & Finance   
3      25-34                     Nonprofits   
4      25-34  Accounting, Banking & Finance   
5      25-34   Education (Higher Education)   
...      ...                            ...   
28101  18-24   Engineering or Manufacturing   
28103  25-34                  Entertainment   
28104  18-24   Engineering or Manufacturing   
28106  35-44                     Nonprofits   
28107  45-54              Computing or Tech   

                                            title title_context  salary  \
0              Research and Instruction Librarian           NaN  55,000   
2                            Marketing Specialist           NaN  34,000   
3                                 Program Manager           NaN  62,000   
4                              Accounting Manager           NaN  60,000   
5                  Scholarly Publishing Librarian           NaN  62,000   
...                                           ...           ...     ...   
28101  Clinical Specialist - Biomedical Engineer            NaN   75000   
28103                               Junior Editor           NaN   72800   
28104                       Applications Engineer           NaN  100000   
28106                  Director of content design           NaN  132000   
28107                                         CMO           NaN  279000   

       additional_comp income_context country           state         city  \
0                  0.0            NaN      us   Massachusetts       Boston   
2                  NaN            NaN      us       Tennessee        Other   
3               3000.0            NaN      us       Wisconsin        Other   
4               7000.0            NaN      us  South Carolina        Other   
5                  NaN            NaN      us   New Hampshire        Other   
...                ...            ...     ...             ...          ...   
28101              NaN            NaN      us        New York        Other   
28103              NaN            NaN      us      California  Los Angeles   
28104          21000.0            NaN      us           Texas        Other   
28106              NaN            NaN      us        Missouri        Other   
28107              0.0            NaN      us      Washington        Other   

           experience field_experience        education gender  \
0           5-7 years        5-7 years  Master's degree  Woman   
2         2 - 4 years      2 - 4 years   College degree  Woman   
3        8 - 10 years        5-7 years   College degree  Woman   
4        8 - 10 years        5-7 years   College degree  Woman   
5        8 - 10 years      2 - 4 years  Master's degree    Man   
...               ...              ...              ...    ...   
28101     2 - 4 years   1 year or less  Master's degree  Woman   
28103       5-7 years        5-7 years  Master's degree  Woman   
28104  1 year or less   1 year or less  Master's degree  Woman   
28106   11 - 20 years    11 - 20 years  Master's degree  Woman   
28107   21 - 30 years    11 - 20 years  Master's degree  Woman   

                                             race  year  month   num_sal  
0                                           White  2021      4   55000.0  
2                                           White  2021      4   34000.0  
3                                           White  2021      4   62000.0  
4                                           White  2021      4   60000.0  
5                                           White  2021      4   62000.0  
...                                           ...   ...    ...       ...  
28101                                       White  2024     12   75000.0  
28103  Hispanic, Latino, or Spanish origin, White  2024     12   72800.0  
28104                                       White  2025      1  100000.0  
28106                                       White  2025      1 

In [1191]:
df.loc[df.num_sal > 1000000]

age                              industry  \
2124        55-64                          Art & Design   
4206        25-34          Engineering or Manufacturing   
5755        25-34                           Health care   
6783        25-34                     Computing or Tech   
9238        35-44  Government and Public Administration   
11134       45-54                                   Law   
15509       35-44                     Computing or Tech   
16165       35-44                                   Law   
17708       25-34                Business or Consulting   
24955       25-34                           Health care   
26558       45-54                Business or Consulting   
28043       25-34                     Computing or Tech   
28084  65 or over                                 Other   

                                                 title  \
2124                                    Owner and CEO    
4206                                 Software Engineer   
5755   Attending Physician (general internal medicine)   
6783                       Principal Software Engineer   
9238                             Senior Policy Advisor   
11134                                          Partner   
15509                                  Product Manager   
16165                                          Partner   
17708                                Senior Consultant   
24955                                Marketing Manager   
26558                                          Partner   
28043                                             Lead   
28084                                  Content creator   

                             title_context     salary  additional_comp  \
2124                                   NaN  3,000,000              NaN   
4206                                   NaN  1,100,000              0.0   
5755                                   NaN  1,900,000              0.0   
6783                                   NaN  1,650,000          25000.0   
9238                                   NaN  1,334,782              0.0   
11134                                  NaN  1,200,000              NaN   
15509                                  NaN  2,111,538          84610.0   
16165                                  NaN  1,100,000              NaN   
17708  Consultant in the IT Strategy field  1,260,000          10000.0   
24955                                  NaN    1250000         350000.0   
26558                  sole proprietor LLC    1300000              0.0   
28043                                  NaN    2600000              NaN   
28084                                  NaN    2000000              NaN   

                                          income_context country  \
2124                                                 NaN      us   
4206                                                 NaN      us   
5755                                                 NaN      us   
6783                                                 NaN      us   
9238   Includes a locality pay for living in a higher...      us   
11134                                                NaN      us   
15509                                                NaN   Other   
16165                                                NaN      us   
17708                                                NaN      us   
24955                                                NaN      us   
26558                                                NaN      us   
28043                                                NaN      us   
28084                                                NaN      us   

                      state     city        experience field_experience  \
2124               New York    Other     21 - 30 years    21 - 30 years   
4206           Pennsylvania    Other         5-7 years        5-7 years   
5755            Connecticut    Other         5-7 years        5-7 years   
6783          Massachusetts   Boston      8 - 10 years        5-7 years   
9238   Di

That looks way more realistic. 

In [1192]:
df['salary'] = df['num_sal'].astype(int)
df.drop('num_sal', axis=1, inplace=True)
df

age                       industry  \
0      25-34   Education (Higher Education)   
2      25-34  Accounting, Banking & Finance   
3      25-34                     Nonprofits   
4      25-34  Accounting, Banking & Finance   
5      25-34   Education (Higher Education)   
...      ...                            ...   
28101  18-24   Engineering or Manufacturing   
28103  25-34                  Entertainment   
28104  18-24   Engineering or Manufacturing   
28106  35-44                     Nonprofits   
28107  45-54              Computing or Tech   

                                            title title_context  salary  \
0              Research and Instruction Librarian           NaN   55000   
2                            Marketing Specialist           NaN   34000   
3                                 Program Manager           NaN   62000   
4                              Accounting Manager           NaN   60000   
5                  Scholarly Publishing Librarian           NaN   62000   
...                                           ...           ...     ...   
28101  Clinical Specialist - Biomedical Engineer            NaN   75000   
28103                               Junior Editor           NaN   72800   
28104                       Applications Engineer           NaN  100000   
28106                  Director of content design           NaN  132000   
28107                                         CMO           NaN  279000   

       additional_comp income_context country           state         city  \
0                  0.0            NaN      us   Massachusetts       Boston   
2                  NaN            NaN      us       Tennessee        Other   
3               3000.0            NaN      us       Wisconsin        Other   
4               7000.0            NaN      us  South Carolina        Other   
5                  NaN            NaN      us   New Hampshire        Other   
...                ...            ...     ...             ...          ...   
28101              NaN            NaN      us        New York        Other   
28103              NaN            NaN      us      California  Los Angeles   
28104          21000.0            NaN      us           Texas        Other   
28106              NaN            NaN      us        Missouri        Other   
28107              0.0            NaN      us      Washington        Other   

           experience field_experience        education gender  \
0           5-7 years        5-7 years  Master's degree  Woman   
2         2 - 4 years      2 - 4 years   College degree  Woman   
3        8 - 10 years        5-7 years   College degree  Woman   
4        8 - 10 years        5-7 years   College degree  Woman   
5        8 - 10 years      2 - 4 years  Master's degree    Man   
...               ...              ...              ...    ...   
28101     2 - 4 years   1 year or less  Master's degree  Woman   
28103       5-7 years        5-7 years  Master's degree  Woman   
28104  1 year or less   1 year or less  Master's degree  Woman   
28106   11 - 20 years    11 - 20 years  Master's degree  Woman   
28107   21 - 30 years    11 - 20 years  Master's degree  Woman   

                                             race  year  month  
0                                           White  2021      4  
2                                           White  2021      4  
3                                           White  2021      4  
4                                           White  2021      4  
5                                           White  2021      4  
...                                           ...   ...    ...  
28101                                       White  2024     12  
28103  Hispanic, Latino, or Spanish origin, White  2024     12  
28104                                       White  2025      1  
28106                                       White  2025      1  
28107          Middle Eastern or Northern African  2025      1  

[23360 rows x 17 columns]

In [1193]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23360 entries, 0 to 28107
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               23360 non-null  category
 1   industry          23360 non-null  category
 2   title             23360 non-null  object  
 3   title_context     6125 non-null   object  
 4   salary            23360 non-null  int64   
 5   additional_comp   17502 non-null  float64 
 6   income_context    2518 non-null   object  
 7   country           23360 non-null  category
 8   state             23360 non-null  category
 9   city              23360 non-null  category
 10  experience        23360 non-null  category
 11  field_experience  23360 non-null  category
 12  education         23204 non-null  category
 13  gender            23218 non-null  category
 14  race              23210 non-null  object  
 15  year              23360 non-null  int32   
 16  month             23360 non

### Title

I'll find common words in titles and add columns for them.

In [1194]:
words = word_tokenize(' '.join(df.title.dropna().str.lower()))
Counter(words).most_common(10)

[('manager', 4461),
 ('senior', 2436),
 ('director', 2406),
 ('engineer', 1865),
 ('analyst', 1411),
 ('assistant', 1403),
 ('of', 1285),
 ('associate', 1177),
 ('software', 1042),
 ('specialist', 1041)]

I'll add columns for the top 3. Without categorizing the column it can't be used in a model. Unless I made a bag of words or tf-idf matrix, but that would add so much complexity to the data. I think it's better to drop it along with title context.

In [1195]:
df['manager'] = df.title.str.contains('manager', case=False)
df['senior'] = df.title.str.contains('senior', case=False)
df['director'] = df.title.str.contains('director', case=False)
df.drop(['title', 'title_context'], axis=1, inplace=True)
df

age                       industry  salary  additional_comp  \
0      25-34   Education (Higher Education)   55000              0.0   
2      25-34  Accounting, Banking & Finance   34000              NaN   
3      25-34                     Nonprofits   62000           3000.0   
4      25-34  Accounting, Banking & Finance   60000           7000.0   
5      25-34   Education (Higher Education)   62000              NaN   
...      ...                            ...     ...              ...   
28101  18-24   Engineering or Manufacturing   75000              NaN   
28103  25-34                  Entertainment   72800              NaN   
28104  18-24   Engineering or Manufacturing  100000          21000.0   
28106  35-44                     Nonprofits  132000              NaN   
28107  45-54              Computing or Tech  279000              0.0   

      income_context country           state         city      experience  \
0                NaN      us   Massachusetts       Boston       5-7 years   
2                NaN      us       Tennessee        Other     2 - 4 years   
3                NaN      us       Wisconsin        Other    8 - 10 years   
4                NaN      us  South Carolina        Other    8 - 10 years   
5                NaN      us   New Hampshire        Other    8 - 10 years   
...              ...     ...             ...          ...             ...   
28101            NaN      us        New York        Other     2 - 4 years   
28103            NaN      us      California  Los Angeles       5-7 years   
28104            NaN      us           Texas        Other  1 year or less   
28106            NaN      us        Missouri        Other   11 - 20 years   
28107            NaN      us      Washington        Other   21 - 30 years   

      field_experience        education gender  \
0            5-7 years  Master's degree  Woman   
2          2 - 4 years   College degree  Woman   
3            5-7 years   College degree  Woman   
4            5-7 years   College degree  Woman   
5          2 - 4 years  Master's degree    Man   
...                ...              ...    ...   
28101   1 year or less  Master's degree  Woman   
28103        5-7 years  Master's degree  Woman   
28104   1 year or less  Master's degree  Woman   
28106    11 - 20 years  Master's degree  Woman   
28107    11 - 20 years  Master's degree  Woman   

                                             race  year  month  manager  \
0                                           White  2021      4    False   
2                                           White  2021      4    False   
3                                           White  2021      4     True   
4                                           White  2021      4     True   
5                                           White  2021      4    False   
...                                           ...   ...    ...      ...   
28101                                       White  2024     12    False   
28103  Hispanic, Latino, or Spanish origin, White  2024     12    False   
28104                                       White  2025      1    False   
28106                                       White  2025      1    False   
28107          Middle Eastern or Northern African  2025      1    False   

       senior  director  
0       False     False  
2       False     False  
3       False     False  
4       False     False  
5       False     False  
...       ...       ...  
28101   False     False  
28103   False     False  
28104   False     False  
28106   False      True  
28107   False     False  

[23360 rows x 18 columns]

In [1196]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23360 entries, 0 to 28107
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               23360 non-null  category
 1   industry          23360 non-null  category
 2   salary            23360 non-null  int64   
 3   additional_comp   17502 non-null  float64 
 4   income_context    2518 non-null   object  
 5   country           23360 non-null  category
 6   state             23360 non-null  category
 7   city              23360 non-null  category
 8   experience        23360 non-null  category
 9   field_experience  23360 non-null  category
 10  education         23204 non-null  category
 11  gender            23218 non-null  category
 12  race              23210 non-null  object  
 13  year              23360 non-null  int32   
 14  month             23360 non-null  int32   
 15  manager           23360 non-null  bool    
 16  senior            23360 non

### Additional Compensation

This one's hard. I'm tempted to add it to the salary column, but that would introduce bias. I could also keep it and use it to predict salary, but there are so many missing values. I guess I kind of get to decide what the purpose of the model is, so I'll say it's to predict salary with additional compensation. I'll add it to salary and drop the column along with it's additional context.

In [1197]:
df.salary.describe()

count    2.336000e+04
mean     9.196248e+04
std      6.812765e+04
min      3.500000e+01
25%      5.616000e+04
50%      7.803900e+04
75%      1.120000e+05
max      3.000000e+06
Name: salary, dtype: float64

In [1198]:
df['additional_comp'] = df.additional_comp.fillna(0)
df['additional_comp'] = df['additional_comp'].astype(int)
df['salary'] = df['salary'] + df['additional_comp']
df.drop(['additional_comp', 'income_context'], axis=1, inplace=True)
df.salary.describe()

count    2.336000e+04
mean     1.014830e+05
std      8.814665e+04
min      3.500000e+01
25%      5.800000e+04
50%      8.100000e+04
75%      1.200000e+05
max      3.000000e+06
Name: salary, dtype: float64

In [1199]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23360 entries, 0 to 28107
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               23360 non-null  category
 1   industry          23360 non-null  category
 2   salary            23360 non-null  int64   
 3   country           23360 non-null  category
 4   state             23360 non-null  category
 5   city              23360 non-null  category
 6   experience        23360 non-null  category
 7   field_experience  23360 non-null  category
 8   education         23204 non-null  category
 9   gender            23218 non-null  category
 10  race              23210 non-null  object  
 11  year              23360 non-null  int32   
 12  month             23360 non-null  int32   
 13  manager           23360 non-null  bool    
 14  senior            23360 non-null  bool    
 15  director          23360 non-null  bool    
dtypes: bool(3), category(9), in

### Race

Like the states, I'll set the rows with multiple chosen races to 'multi'.

In [1200]:
race_options = ['Asian or Asian American', 'Black or African American', 'Hispanic, Latino, or Spanish origin', 'Middle Eastern or Northern African', 'Native American or Alaska Native', 'White', 'Another option not listed here or prefer not to answer']
df.loc[~df.race.isin(race_options), 'race'] = 'multi'
df['race'] = df.race.astype('category')
df.race.value_counts()

race
White                                                     19259
multi                                                      1244
Asian or Asian American                                    1132
Black or African American                                   620
Hispanic, Latino, or Spanish origin                         548
Another option not listed here or prefer not to answer      466
Middle Eastern or Northern African                           55
Native American or Alaska Native                             36
Name: count, dtype: int64

In [1201]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23360 entries, 0 to 28107
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   age               23360 non-null  category
 1   industry          23360 non-null  category
 2   salary            23360 non-null  int64   
 3   country           23360 non-null  category
 4   state             23360 non-null  category
 5   city              23360 non-null  category
 6   experience        23360 non-null  category
 7   field_experience  23360 non-null  category
 8   education         23204 non-null  category
 9   gender            23218 non-null  category
 10  race              23360 non-null  category
 11  year              23360 non-null  int32   
 12  month             23360 non-null  int32   
 13  manager           23360 non-null  bool    
 14  senior            23360 non-null  bool    
 15  director          23360 non-null  bool    
dtypes: bool(3), category(10), i

Woot! No more object columns. All that's left is missing values.

In [1202]:
df.isnull().sum()

age                   0
industry              0
salary                0
country               0
state                 0
city                  0
experience            0
field_experience      0
education           156
gender              142
race                  0
year                  0
month                 0
manager               0
senior                0
director              0
dtype: int64

The gender question has a 'prefer not to answer' option. That's what the nulls will be set to.

In [1203]:
df['gender'] = df.gender.fillna('Other or prefer not to answer')
df.gender.value_counts()

gender
Woman                            18056
Man                               4326
Non-binary                         598
Other or prefer not to answer      379
Prefer not to answer                 1
Name: count, dtype: int64

How? How did that get entered?

In [1204]:
df.loc[df.gender == 'Prefer not to answer']

age                 industry  salary country     state   city  \
6477  35-44  Agriculture or Forestry   88010      us  Nebraska  Other   

          experience field_experience education                gender  \
6477  1 year or less        5-7 years       PhD  Prefer not to answer   

                                                   race  year  month  manager  \
6477  Another option not listed here or prefer not t...  2021      4    False   

      senior  director  
6477   False     False

Whatever

In [1205]:
df.loc[df.gender == 'Prefer not to answer', 'gender'] = 'Other or prefer not to answer'
df['gender'] = df.gender.cat.remove_unused_categories()
df.gender.value_counts()

gender
Woman                            18056
Man                               4326
Non-binary                         598
Other or prefer not to answer      380
Name: count, dtype: int64

Wow, way more women than men.

## Reflection
Write a short reflection (400-500 words) answering the following: 
- What were the biggest issues you encountered in the messy dataset?
- How did cleaning the dataset improve its usability for machine learning?
- What would happen if we trained a model on the messy dataset vs. the cleaned one?
- Do you feel you skewed or biased the dataset while cleaning it?

[Add you answer to this markdown cell]

## Deliverables
Upload your Jupyter Notebook to your GitHub repo and then provide a link to that repo in Worlclass. 